mergify_engine/worker.py

```diff
@@ -0,0 +1,366 @@
+                },
+                use_bin_type=True,
+            ),
+        },
+    )
+    # NOTE(sileht): Add pull request stream to process to the list, only if it
+    # does not exists, to not update the score(date)
+    transaction.zadd("streams", {stream_name: score}, nx=True)
+    transaction.execute()
+    LOG.info(
+        "pushed to worker",
+        gh_owner=owner,
+        gh_repo=repo,
+        gh_pull=pull_number,
+        event_type=event_type,
+    )
+
+
+def run_engine(installation_id, owner, repo, pull_number, sources):
+    try:
+        installation = github.get_installation(owner, repo, installation_id)
+    except exceptions.MergifyNotInstalled:
+        return
+    with github.get_client(owner, repo, installation) as client:
+        pull = client.item(f"pulls/{pull_number}")
+        engine.run(client, pull, sources)
+
+
+class StreamProcessor:
+    def __init__(self, worker_count):
+        self._worker_count = worker_count
+        self._loop = asyncio.get_running_loop()
+        self._executor = concurrent.futures.ThreadPoolExecutor(max_workers=worker_count)
+        self._pending_streams = set()
+        self._redis = None
+
+    async def connect(self):
+        self._redis = await utils.create_aioredis_for_stream()
+
+    async def close(self):
+        if self._redis:
+            self._redis.close()
+            await self._redis.wait_closed()
+            self._redis = None
+
+    @contextlib.asynccontextmanager
+    async def next(self):
+        # TODO(sileht): We can cache locally the result as the order is not going to
+        # change, and if it change we don't care
+        now = time.time()
+        streams = [
+            s
+            for s in await self._redis.zrangebyscore(
+                "streams", max=now, offset=0, count=self._worker_count * 2
+            )
+            if s not in self._pending_streams
+        ]
+        if streams:
+            self._pending_streams.add(streams[0])
+            try:
+                yield streams[0].decode()
+            finally:
+                self._pending_streams.remove(streams[0])
+        else:
+            yield
+
+    async def _inc_and_get_attempts(self, key):
+        transaction = self._redis.multi_exec()
+        transaction.hincrby("attempts", key)
+        transaction.hget("attempts", key)
+        return (await transaction.execute())[0]
+
+    async def _run_engine(self, installation_id, owner, repo, pull_number, sources):
+        try:
+            await self._loop.run_in_executor(
+                self._executor,
+                functools.partial(
+                    run_engine, installation_id, owner, repo, pull_number, sources,
+                ),
+            )
+        # Translate in more understandable exception
+        except exceptions.MergeableStateUnknown as e:
+            key = f"pull~{installation_id}~{owner}~{repo}~{pull_number}"
+            attempts = await self._inc_and_get_attempts(key)
```

Comment:
I'd say maybe compute the key string in `_inc_and_get_attemps(installation_id, owner, repo, pull_number)`?

Reply:
We can't because we have two kind of key pull~XXX ans stream~XXX